Userland/Libraries/LibWeb/FileAPI/FileReader.cpp

```diff
@@ -0,0 +1,359 @@
+            Platform::EventLoopPlugin::the().spin_until([&]() {
+                return chunk_promise->state() == JS::Promise::State::Fulfilled || chunk_promise->state() == JS::Promise::State::Rejected;
+            });
+
+            // 2. If chunkPromise is fulfilled, and isFirstChunk is true, queue a task to fire a progress event called loadstart at fr.
+            // NOTE: ISSUE 2 We might change loadstart to be dispatched synchronously, to align with XMLHttpRequest behavior. [Issue #119]
+            if (chunk_promise->state() == JS::Promise::State::Fulfilled && is_first_chunk) {
+                HTML::queue_global_task(HTML::Task::Source::FileReading, realm.global_object(), [this, &realm]() {
+                    dispatch_event(DOM::Event::create(realm, HTML::EventNames::loadstart));
+                });
+            }
+
+            // 3. Set isFirstChunk to false.
+            is_first_chunk = false;
+
+            VERIFY(chunk_promise->result().is_object());
+            auto& result = chunk_promise->result().as_object();
+
+            auto value = MUST(result.get(vm.names.value));
+            auto done = MUST(result.get(vm.names.done));
+
+            // 4. If chunkPromise is fulfilled with an object whose done property is false and whose value property is a Uint8Array object, run these steps:
+            if (chunk_promise->state() == JS::Promise::State::Fulfilled && !done.as_bool() && is<JS::Uint8Array>(value.as_object())) {
+                // 1. Let bs be the byte sequence represented by the Uint8Array object.
+                auto const& byte_sequence = verify_cast<JS::Uint8Array>(value.as_object());
+
+                // 2. Append bs to bytes.
+                bytes.append(byte_sequence.data());
+
+                // FIXME: 3. If roughly 50ms have passed since these steps were last invoked, queue a task to fire a progress event called progress at fr.
+
+                // 4. Set chunkPromise to the result of reading a chunk from stream with reader.
+                chunk_promise = MUST(reader->read());
+            }
+            // 5. Otherwise, if chunkPromise is fulfilled with an object whose done property is true, queue a task to run the following steps and abort this algorithm:
+            else if (chunk_promise->state() == JS::Promise::State::Fulfilled && done.as_bool()) {
+                HTML::queue_global_task(HTML::Task::Source::FileReading, realm.global_object(), [this, bytes, type, &realm, encoding_name, blobs_type]() {
+                    // 1. Set fr’s state to "done".
+                    m_state = State::Done;
+
+                    // 2. Let result be the result of package data given bytes, type, blob’s type, and encodingName.
+                    auto result = blob_package_data(realm, bytes, type, blobs_type, encoding_name);
+
+                    // 3. If package data threw an exception error:
+                    if (result.is_error()) {
+                        // FIXME: 1. Set fr’s error to error.
```

Comment:
I had a feeling there was a way to convert the error types, but I can't find it...

Reply:
I believe the closest we have to this right now is: https://github.com/SerenityOS/serenity/blob/b0c1dc14f39bdafd35ba50f28499e480fa4d2d12/Userland/Libraries/LibWeb/Bindings/ExceptionOrUtils.h#L61

Which is the opposite to what I want unfortunately